### PR TITLE
Fixes fire on create issue with watches

### PR DESCRIPTION
**Pull request by Sudhar287**
For #571 

Consul watches fire when it is first created. This poses a problem for all the users who don't want the `fire on create` behavior. Some users have 100s of changes taking place and they don't want newly defined watches to fire after creating because this utilizes/holds resources. 

The fix should essentially solve the watches fire on create issue and also give the users the power to set/unset this behavior.

### CODE REVIEW

File: api/watch/plan.go [hunk 1]
```diff
@@ -118,13 +119,21 @@ OUTER:
 		p.lastResult = result
 		// If a hybrid handler exists use that
 		if p.HybridHandler != nil {
+			if first == true && p.FireOnCreate == "no" {
```
Comment (Sudhar287) [inline api/watch/plan.go:122]: Trying to be explicit with `first == true`. Please let me know if you want it changed. :)

File: api/watch/watch.go [hunk 2]
```diff
@@ -171,6 +173,9 @@ func ParseExempt(params map[string]interface{}, exempt []string) (*Plan, error)
 	case "script":
 		// Let the caller check for configuration in exempt parameters
 	}
+	if err := assignValueBool(params, "ConsulReloadTriggersWatch", &plan.ConsulReloadTriggersWatch); err != nil {
```
Comment (mknapphrt) [inline api/watch/watch.go:176]: Seems that all the other params are stored as lowercase, snakecase variables.

**Comment by Sudhar287** [comment]
Thank you for the comment and code review @mknapphrt! 

1. Yes, this solution will work only while using `consul agent watch` for eg: `consul agent -dev -hcl 'watches {type = "key", key = "/test", handler_type = "script", args = ["sh", "-c", "echo test >> /tmp/consulwatch"] }'`. [This](https://github.com/hashicorp/consul/issues/571#issuecomment-612920278) comment gives more insight into the design decision. (And while you're at it, please do consider reading the most recent comments to get more insights into my design choices). Watches declared using `consul watch` are not tied to the agent lifecycle and are not affected even by `consul reload`. I figured that `consul watch` should not be coupled with the `fire-on-create` issue. But I'm ready to change the implementation if the maintainers have other ideas! :)
2. `fire-on-create` is a consequence of reloading using `consul reload`. This issue can also be looked at as a `fire-on-reload` problem.

The above are the design decisions I took after discussing on the thread linked above. Awaiting feedback from the maintainers on the parts of the implementation that should be modified. This PR should also be accompanied by a modification to the documentation which I'll do once my code is close to getting approved.

File: agent/config/builder.go [hunk 3]
```diff
@@ -987,6 +987,7 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
 		VerifyOutgoing:                         verifyOutgoing,
 		VerifyServerHostname:                   verifyServerName,
 		Watches:                                c.Watches,
+		ConsulReloadTriggersWatch:              b.stringValWithDefault(c.FireWatchesOnCreate, "yes"),
```
[the remark this answers is not in the document]
Comment (Sudhar287) [reply agent/config/builder.go:990]: I though a `yes`/`no` `string` would be more intuitive while using the app from the command line...

**Comment by kaspergrubbe** [comment]
Thank you so much for your progress on this @Sudhar287 🎉

**Comment by sc0rp10** [comment]
@ Sudhar287 am i right that there will be different behaviors between console-call consul watch and watches section placed into the config file?

**Review by dnephin — COMMENTED** [review]
Thank you for working on this PR! Sorry it has taken us such a long time to respond.

I just came across some notes that we had taken a few months ago, and it looks like we forgot to reply to this PR. It has been a while, so I apologize if these notes are not clear. If you are still interested in working on this PR I can take another detailed look.

From our notes, we were thinking this new setting would unfortunately not be safe for the config reload case. Any changes made during config reload would not be noticed by any watchers because this setting would prevent the notification.

In `api/watch/plan.go`, just above where the changes were made, it seems like there is some logic to copy and compare the new value and result to the previous value and result. It attempts to skip the fire if they match. If I understand this correctly, that code should already be handling this case. It seems like maybe there is a bug in that comparison, which is what is causing this problem.

I think if we can find that bug we can probably fix the issue without a new config setting.

**Comment by jsosulska** [comment]
Closing as there has been no action on this PR. As @dnephin pointed out, the implementation details will be different, and this PR has drifted. Closing. Track https://github.com/hashicorp/consul/issues/571 for changes.

**Pull request closed without merging**